atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/store/AtlasDbTransactionStore.java

```diff
@@ -0,0 +1,159 @@
+        } catch (SafeIllegalArgumentException e) {
+            throw e;
+        } catch (Exception e) {
+            log.info("Failed to record transaction due to an exception", e);
+            return Optional.empty();
+        }
+    }
+
+    private TableReference getTableReferenceOrThrow(TransactionAction action) {
+        return getTableReferenceOrThrow(action.table());
+    }
+
+    private TableReference getTableReferenceOrThrow(String table) {
+        return Optional.ofNullable(tables.get(table))
+                .orElseThrow(() -> new SafeIllegalArgumentException(
+                        "Transaction action has unknown table.",
+                        SafeArg.of("readTransactionTableName", table),
+                        SafeArg.of("availableTables", tables)));
+    }
+
+    private class AtlasDbTransactionActionVisitor implements TransactionActionVisitor<WitnessedTransactionAction> {
+
+        private final Transaction transaction;
+
+        public AtlasDbTransactionActionVisitor(Transaction transaction) {
+            this.transaction = transaction;
+        }
+
+        @Override
+        public WitnessedTransactionAction visit(ReadTransactionAction readTransactionAction) {
+            Cell cell = AtlasDbUtils.toAtlasCell(readTransactionAction.cell());
+            TableReference tableReference = getTableReferenceOrThrow(readTransactionAction);
+            Map<Cell, byte[]> cells =
+                    transaction.get(tableReference, Set.of(AtlasDbUtils.toAtlasCell(readTransactionAction.cell())));
+            Optional<Integer> value = Optional.ofNullable(cells.get(cell)).map(Ints::fromByteArray);
+            return readTransactionAction.witness(value);
+        }
+
+        @Override
+        public WitnessedTransactionAction visit(WriteTransactionAction writeTransactionAction) {
+            TableReference tableReference = getTableReferenceOrThrow(writeTransactionAction);
+            transaction.put(
+                    tableReference,
+                    Map.of(
+                            AtlasDbUtils.toAtlasCell(writeTransactionAction.cell()),
+                            Ints.toByteArray(writeTransactionAction.value())));
+            return writeTransactionAction.witness();
+        }
+
+        @Override
+        public WitnessedTransactionAction visit(DeleteTransactionAction deleteTransactionAction) {
+            TableReference tableReference = getTableReferenceOrThrow(deleteTransactionAction);
+            transaction.delete(tableReference, Set.of(AtlasDbUtils.toAtlasCell(deleteTransactionAction.cell())));
+            return deleteTransactionAction.witness();
+        }
+    }
+
+    public static AtlasDbTransactionStore create(
+            TransactionManager transactionManager, Map<TableReference, byte[]> tables) {
+        EntryStream.of(tables)
+                .forKeyValue((table, metadata) ->
+                        transactionManager.getKeyValueService().createTable(table, metadata));
```

Comment:
This is just `transactionManager.getKeyValueService().createTables(tables)` I think

Reply:
lol